scope.go

```diff
@@ -0,0 +1,622 @@
+var _ network.TransactionalScope = (*resourceScope)(nil)
+
+func newResourceScope(limit Limit, constraints []*resourceScope, name string) *resourceScope {
+	for _, cst := range constraints {
+		cst.IncRef()
+	}
+	return &resourceScope{
+		rc:          resources{limit: limit},
+		constraints: constraints,
+		name:        name,
+	}
+}
+
+func newTxnResourceScope(owner *resourceScope) *resourceScope {
+	return &resourceScope{
+		rc:    resources{limit: owner.rc.limit},
+		owner: owner,
+		name:  fmt.Sprintf("%s.txn", owner.name),
+	}
+}
+
+// Resources implementation
+func (rc *resources) checkMemory(rsvp int64, prio uint8) error {
+	// overflow check; this also has the side effect that we cannot reserve negative memory.
+	newmem := rc.memory + rsvp
+	limit := rc.limit.GetMemoryLimit()
+	threshold := (1 + int64(prio)) * limit / 256
+
+	if newmem > threshold {
+		return network.ErrResourceLimitExceeded
+	}
+
+	return nil
+}
+
+func (rc *resources) reserveMemory(size int64, prio uint8) error {
+	if err := rc.checkMemory(size, prio); err != nil {
+		return err
+	}
+
+	rc.memory += size
+	return nil
+}
+
+func (rc *resources) releaseMemory(size int64) {
+	rc.memory -= size
+
+	// sanity check for bugs upstream
+	if rc.memory < 0 {
+		panic("BUG: too much memory released")
+	}
+}
+
+func (rc *resources) addStream(dir network.Direction) error {
+	if dir == network.DirInbound {
+		return rc.addStreams(1, 0)
+	}
+	return rc.addStreams(0, 1)
+}
+
+func (rc *resources) addStreams(incount, outcount int) error {
+	if incount > 0 && rc.nstreamsIn+incount > rc.limit.GetStreamLimit(network.DirInbound) {
+		return fmt.Errorf("cannot reserve stream: %w", network.ErrResourceLimitExceeded)
+	}
+	if outcount > 0 && rc.nstreamsOut+outcount > rc.limit.GetStreamLimit(network.DirOutbound) {
+		return fmt.Errorf("cannot reserve stream: %w", network.ErrResourceLimitExceeded)
+	}
+
+	rc.nstreamsIn += incount
+	rc.nstreamsOut += outcount
+	return nil
+}
+
+func (rc *resources) removeStream(dir network.Direction) {
+	if dir == network.DirInbound {
+		rc.removeStreams(1, 0)
+	} else {
+		rc.removeStreams(0, 1)
+	}
+}
+
+func (rc *resources) removeStreams(incount, outcount int) {
+	rc.nstreamsIn -= incount
+	rc.nstreamsOut -= outcount
+
+	if rc.nstreamsIn < 0 || rc.nstreamsOut < 0 {
+		panic("BUG: too many streams released")
+	}
+}
+
+func (rc *resources) addConn(dir network.Direction, usefd bool) error {
+	var fd int
+	if usefd {
+		fd = 1
+	}
+
+	if dir == network.DirInbound {
+		return rc.addConns(1, 0, fd)
+	}
+
+	return rc.addConns(0, 1, fd)
+}
+
+func (rc *resources) addConns(incount, outcount, fdcount int) error {
+	if incount > 0 && rc.nconnsIn+incount > rc.limit.GetConnLimit(network.DirInbound) {
+		return fmt.Errorf("cannot reserve connection: %w", network.ErrResourceLimitExceeded)
+	}
+	if outcount > 0 && rc.nconnsOut+outcount > rc.limit.GetConnLimit(network.DirOutbound) {
+		return fmt.Errorf("cannot reserve connection: %w", network.ErrResourceLimitExceeded)
+	}
```

Comment:
we dont need to be separate for this i think,  and it could lead to starvation.
Better to explicitly specify in/out i think.

Having said that, if you feel strongly about it, we could have an aggregate limit (for streams too).

Reply:
as we discussed in sync, added a total limit for both conns and streams.